### ISSUE TITLE
Hadolint checkstyle report unexpectedly hanging when Jenkins Plugin records issues

### ISSUE THREAD

**Issue by donalthurley** [hadolint/hadolint#529]
- [x] This is a bug report
- [x] This is a feature request
- [ ] I searched existing issues before opening this one

### Expected behavior

I'm not sure if this is a bug or a feature request.

We are using the docker version of hadolint to run checks on our Dockerfiles in our Jenkins pipelines.
We specify the `checkfile` format when running and we save the results to a report file.
We then use the [Jenkins Warnings Ng Plugin](https://www.jenkins.io/doc/pipeline/steps/warnings-ng/) to record the issues and to publish the report to Jenkins.

We would have expected this to work fine using a command like `docker run --rm -i hadolint/hadolint hadolint -f checkstyle - < Dockerfile > report.xml`. 

### Actual behavior

We found that when we called the Jenkins `recordIssues` function with the saved report file that the pipeline hung on this call.

It turns out that it hangs because in the `checkstyle` report there is a tag `<file name='/dev/stdin' >` and the Jenkins plugin tries to access this file and so waits to read it.

Maybe hadolint could remove the `name` attribute on the `<file>` tag if the input has been piped in and it is not coming from a real file?
That is how we have fixed our problem for the moment by taking away the `name` attribute on that tag using `sed`.

### Steps to reproduce the behavior

docker run --rm -i hadolint/hadolint hadolint -f checkstyle - < Dockerfile

The XML generated is as follows

`<?xml version='1.0' encoding='UTF-8'?><checkstyle version='4.3'><file name='/dev/stdin' ><error line='1' column='1' severity='warning' message='Using latest is prone to errors if the image will ever update. Pin the version explicitly to a release tag' source='DL3007' /></file></checkstyle>`

In our Jenkins pipelines we run the following command and save it to a report file.

docker run --rm -i hadolint/hadolint hadolint -f checkstyle - < Dockerfile > report.xml

We then call `recordIssues` on the [Jenkins Warnings Ng Plugin](https://www.jenkins.io/doc/pipeline/steps/warnings-ng/)

`recordIssues(tools: [checkStyle(pattern: 'report.xml')])`

The recordIssues seems to hang but is blocking waiting for input.

**Output of `hadolint --version` or
  `docker run --rm hadolint/hadolint hadolint --version` or
  `docker run --rm ghcr.io/hadolint/hadolint hadolint --version`:**

```bash
Haskell Dockerfile Linter v1.21.0-0-gcb19447
```

**Dockerfile** (if relevant)

```Dockerfile
FROM ubuntu:latest
```

**Additional environment details (OS, stack version, etc.)**

**Comment by lorenzo** [comment]
I think removing the file name from the report makes sense, would you like to submit a pull request with the fix?

**Comment by donalthurley** [comment]
Just thinking about it again I could do a PR for removing the file attribute from [Checkstyle.hs](https://github.com/hadolint/hadolint/blob/master/src/Hadolint/Formatter/Checkstyle.hs) but not sure now if that is the best solution.

I think it is a good feature to have the file attribute when running Hadolint from the command line because the Jenkins Plugin can then link the line errors to the location in the file.

However if you are running Hadolint docker then it's a problem defining what the path to that file is.
If you pipe the Dockerfile then the file path in the attribute is set to `<file name='/dev/stdin' >` which causes the hanging problem.

If you try to mount the local directory inside the container I'm not sure if you can get the correct file path.
For instance, if your Dockerfile is in the current local path you want to get this result `<file name='Dockerfile' >`.
But `docker run --rm -i -v "$(pwd)":/work hadolint/hadolint hadolint -f checkstyle /work/Dockerfile` will give this result `<file name='/work/Dockerfile' >`.
If you have the wrong file path the Jenkins Plugin will still work and it will shows the errors. It just can't link the file with the errors.

Our fix was using `sed` which could be used to put in the correct name after it was run in docker.
For insatnce, `docker run --rm -i hadolint/hadolint hadolint -f checkstyle - < Dockerfile > report.xml; sed -iE "s/<file name='[^']*'/<file name='Dockerfile'/" report.xml`

I'm still unsure what is the right solution for Hadolint.

1. Should it be just a warning in the Hadolint docker documentation when using the checkstyle format that you may have problems with the generated file path and maybe provide some example solutions.
2. Or should the file attribute be removed altogether. But then you lose the linking with the file and the errors which works fine for Hadolint on the command line.

**Comment by lorenzo** [comment]
maybe we need an option for specifying the file name in the report?

**Comment by donalthurley** [comment]
I think that would be a better solution.

There could be an option added like `--local-file-path`.
A local help message like `"The local file path. If present this value will be set as the local path of the input source file in the generated report"`

This option would only be used when the output format `-f, --format` option was also being used.
It would be needed I think by all supported formats.
It also wouldn't make sense to use this option if you input from multiple docker files as you can only redirect the output to one report file.

I've never programmed in Haskell before so I'm not sure if I can figure how to do this if you want to make this change.
I can give it a try if you think it's not too hard and try and submit a PR?

**Comment by lorenzo** [comment]
@donalthurley sorry for the delay in responding to this. I think this would be simple enough to implement, if you want to give it a try. The trick would be to check whether or not the file name is stdin here: https://github.com/hadolint/hadolint/blob/master/src/Hadolint/Formatter/Checkstyle.hs#L97 and change it to something safe.

le time know if you need help getting started with this.

**Comment by donalthurley** [comment]
@lorenzo sorry for my delay, I had a look at this again and thought maybe a command option with the name `--file-path-in-report` might be clearer and submitted changes for that. I'm not sure, it still might be a little confusing.

**Comment by m-ildefons** [comment]
Seems like this issue has been fixed by #642, right? I'll close this.
